### PR TITLE
K-Stereographic (revised)

### PR DESCRIPTION
**Pull request by ferrine**
Due to rebase conflicts and some design concerns in #123 we (with @andbloch) considered to do this PR 2 folded.

1) implementation
2) docs

both aimed to develop branch. After it is finished, we merge to master

### CODE REVIEW

File: geoopt/manifolds/stereographic/math.py [hunk 1]
```diff
@@ -0,0 +1,2018 @@
...
+        return (
+            x
+            + 1 / 3 * k * x ** 3
+            + 2 / 15 * k ** 2 * x ** 5
+            + 17 / 315 * k ** 3 * x ** 7
+            + 62 / 2835 * k ** 4 * x ** 9
+        )
+    else:
+        raise RuntimeError("order not in [-1, 5]")
+
```
Comment (rrkarim) [inline geoopt/manifolds/stereographic/math.py]: Can loop be use here to simplify the code?

Comment (ferrine) [reply geoopt/manifolds/stereographic/math.py]: I thought about it, don't now how much readability it adds

Comment (rrkarim) [reply geoopt/manifolds/stereographic/math.py]: Well maybe dynamic order with numerator/denominator ([A036279](https://oeis.org/A036279)/[A002430](https://oeis.org/A002430) ) sequence in list somewhere in globals?

File: geoopt/manifolds/stereographic/math.py [hunk 2]
```diff
@@ -0,0 +1,2018 @@
...
+    -----
+    This implementation uses stereographic projection
+    """
+    return _antipode(x, k, dim=dim)
+
+
+# Max: progress for the commit stops here
+
+
+def antipode(x: torch.Tensor, *, k: torch.Tensor, dim: int = -1):
```
Comment (rrkarim) [inline geoopt/manifolds/stereographic/math.py]: Repeated definition

File: geoopt/manifolds/stereographic/math.py [hunk 3]
```diff
@@ -0,0 +1,2018 @@
...
+    -------
+    tensor
+        the result of the projection
+    """
+    return _inv_sproj(x, k, dim=dim)
+
+
+@torch.jit.script
+def _inv_sproj(x: torch.Tensor, k: torch.Tensor, dim: int = -1):
+    # TODO: test
```
Comment (rrkarim) [inline geoopt/manifolds/stereographic/math.py]: There are tests for this op.

File: geoopt/manifolds/stereographic/math.py [hunk 4]
```diff
@@ -0,0 +1,2018 @@
...
+    -------
+    tensor
+        the result of the projection
+    """
+    return _sproj(x, k, dim=dim)
+
+
+@torch.jit.script
+def _sproj(x: torch.Tensor, k: torch.Tensor, dim: int = -1):
+    # TODO: test
```
Comment (rrkarim) [inline geoopt/manifolds/stereographic/math.py]: here too.

File: geoopt/manifolds/stereographic/math.py [hunk 5]
```diff
@@ -0,0 +1,2018 @@
...
+    tensor
+    """
+    return _parallel_transport0back(x, v, k=k, dim=dim)
+
+
+@torch.jit.script
+def _parallel_transport0back(
+    x: torch.Tensor, v: torch.Tensor, k: torch.Tensor, dim: int = -1
+):
+    return v / (1 + k * x.pow(2).sum(dim=dim, keepdim=True)).clamp_min(1e-15)
```
Comment (rrkarim) [inline geoopt/manifolds/stereographic/math.py]: Maybe fuse here.

Comment (ferrine) [reply geoopt/manifolds/stereographic/math.py]: is there anything to fuse?

File: geoopt/manifolds/stereographic/math.py (hunk 4 above)
Comment (ferrine) [reply geoopt/manifolds/stereographic/math.py]: done

File: geoopt/manifolds/stereographic/math.py (hunk 3 above)
Comment (ferrine) [reply geoopt/manifolds/stereographic/math.py]: done

File: geoopt/manifolds/stereographic/math.py (hunk 2 above)
Comment (ferrine) [reply geoopt/manifolds/stereographic/math.py]: done

File: geoopt/manifolds/stereographic/math.py (hunk 1 above)
Comment (ferrine) [reply geoopt/manifolds/stereographic/math.py]: I consider rewriting this to Cpp at some point. I'd prefer to leave this for now without overcomplication.

Comment (ferrine) [reply geoopt/manifolds/stereographic/math.py]: And I don't think it is important to have a very long tailor series

File: geoopt/manifolds/stereographic/math.py (hunk 5 above)
Comment (rrkarim) [reply geoopt/manifolds/stereographic/math.py]: `addcmul` instead of `1 + a * b`

Comment (ferrine) [reply geoopt/manifolds/stereographic/math.py]: https://github.com/pytorch/pytorch/issues/23655

Comment (ferrine) [reply geoopt/manifolds/stereographic/math.py]: I see torch script should do this for me, no need to change

Comment (ferrine) [reply geoopt/manifolds/stereographic/math.py]: I'm not sure if it actually fuses ops. Inspecting the resulting graph I do not see any fused ops

Comment (rrkarim) [reply geoopt/manifolds/stereographic/math.py]: I always forget that all the ops are serialized now. Though not sure how fusing handled by torchscript. Anyay, resolved.

**Comment by ferrine** [comment]
I have nans using this code. I want now more tests verifying grads are not none

File: docs/extended/stereographic.rst [hunk 6]
```diff
@@ -0,0 +1,146 @@
+:math:`\kappa`-Stereographic Projection model
+=============================================
+
+Stereographic projection models comes to bind constant curvature spaces. Such as spheres,
+hyperboloids and regular Euclidean manifold. Let' look at what does this mean. As we mention
```
Comment (rrkarim) [inline docs/extended/stereographic.rst]: Let's

Comment (rrkarim) [reply docs/extended/stereographic.rst]: mentioned

File: docs/extended/stereographic.rst [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+   Grid of Geodesics for :math:`\kappa=-1`, credits to `Andreas Bloch`_
+
+First of all we note, that Poincare ball is embedded in a Sphere of radius :math:`r=1/\sqrt{\kappa}`,
+where c is negative curvature. We also note, as :math:`\kappa` goes to :math:`0`, we recover infinite radius ball.
+We should expect this limiting behaviour recovers Euclidean geometry.
+
+Spherical Spaces
+----------------
+Another case of constant curvature manifolds is sphere. Unlike Hyperboloid this manifold is compact and has
+:math:`\kappa`. But still we can embed a sphere onto a plane ignoring one of the poles.
```
Comment (rrkarim) [inline docs/extended/stereographic.rst]: \kappa what?
